### ISSUE TITLE
cmd/go: go test -fullpath should include full paths for build errors too

### ISSUE THREAD

**Issue by rogpeppe** [golang/go#60451]
### What version of Go are you using (`go version`)?

<pre>
f90b4cd6554f4f20280aa5229cf42650ed47221d
</pre>

### Does this issue reproduce with the latest release?

N/A

### What did you do?

I ran this testscript:

```
! exec go test -C ./x -fullpath .
! stderr '\./x.go'

-- x/x.go --
package foo

var x = mistake

-- go.mod --
module test
```

### What did you expect to see?

A passing test.

### What did you see instead?

```
> ! exec go test -C ./x -fullpath .
[stdout]
FAIL
[stderr]
# test/x
./x.go:3:9: undefined: mistake
[exit status 1]
> ! stderr '\./x.go'
FAIL: /tmp/testscript142857279/x.txtar/script.txtar:2: unexpected match for `\./x.go` found in stderr: ./x.go
```

It seems to me that the intent of `-fullpath` is to cause all filenames to print in a way that's absolute, even when those errors come from the compiler rather than the test binary. I'd expect the compiler error to print an absolute path there too.

**Comment by rogpeppe** [comment]
Yup, that sounds about right. I think that would be worth doing.

**Comment by bcmills** [comment]
Sounds reasonable. This is essentially a revision on proposal #37708, so it should probably be a proposal itself.

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://go.dev/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://go.dev/s/proposal-status#likely-accept)**.
— rsc for the proposal review group

**Comment by rsc** [comment]
No change in consensus, so **[accepted](https://go.dev/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— rsc for the proposal review group

**Comment by quantonganh** [comment]
If we add `-fullpath` as build option, it may lead to confusion with the `-trimpath` flag:

```
        -trimpath
                remove all file system paths from the resulting executable.
                Instead of absolute file system paths, the recorded file names
                will begin either a module path@version (when using modules),
                or a plain import path (when using the standard library, or GOPATH).
```

What do you think if we just pass `-test.fullpath` to cmd/compile?

**Comment by bcmills** [comment]
@quantonganh, if `-fullpath` and `-trimpath` are both specified, the “full path” logged by the test should correspond to the short filename joined with the package import path.

It isn't clear to me what `-fullpath` as a build option should do in that case, though. 🤔

**Comment by bcmills** [comment]
Also: if `-fullpath` is (also) a build flag, should it apply uniformly to all build commands? It seems inconsistent to apply it to only the build portion of `go test`.

**Comment by karnigen** [comment]
I'm glad it's implemented now, but I can't figure out how to use `-fullpath` during builds, not just tests? What should the command look like specifically?
I'm trying, for example: `go -C go build -o golib.a -fullpath -buildmode=c-archive golib.go`, but I can't get it to work. How is it supposed to be used?

**Comment by karnigen** [comment]
To streamline editor error navigation, I've created a simple filter as a workaround. This tool reads stdin, replaces relative paths, and writes the modified output to stdout.
Usage:` go -C <project_path> build . 2>&1 | gobuildfilter [project_path]`

```
package main

func usage() {
	fmt.Println("Usage: go -C <project_path> build . 2>&1 | gobuildfilter [project_path]")
	fmt.Println("       filter replaces './' with the specified path, tu support go -C and editor integration")
	fmt.Println("       In bash script 'set -eo pipefail' or 'bash -eo pipefail' to stop on error")
	fmt.Println()
	flag.PrintDefaults()
}

func main() {
	flag.Usage = usage
	flag.Parse()

	// Reads stdin, replaces "./", writes to stdout.
	path := "."
	if flag.NArg() == 1 {
		path = flag.Arg(0)
	}

	if path[len(path)-1] != '/' {
		path += "/"
	}

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		fmt.Println(strings.Replace(line, "./", path, -1))
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
	}
}

```

**Comment by matloob** [comment]
@karnigen This isn't implemented in the go command yet. There's a CL out for review but it hasn't been submitted or released.

**Comment by firelizzard18** [comment]
As a tool builder this would be very useful. Now that `go test -json` outputs build errors as a test event it is trivial to capture them. Updating those events so they respect -fullpath would make it much easier to attach the error to the correct location when reporting it to the development environment.
